### DISCUSSION TITLE
Support for system-versioned tables

### DISCUSSION THREAD

**Discussion by gnom7** [querydsl/querydsl#2930] [Q&A]
Hey
I would like to run queries as described [here](https://mariadb.com/kb/en/system-versioned-tables/) but cannot find a way to build such a query.
Is there support for **system-versioned tables** (SQL:2011 standard)? If not, is there any workaround one could resort to?

I tried to use native sql sub query to only specify table version:
```
SQLQueryFactory factory = new SQLQueryFactory(configuration, dataSource);
List<String> names = factory
        .select(new QEntity("entity_slice").entityName)
        .from(Expressions.template(QEntity.class, "(select * from entity_table for system_time as of transaction 12345) entity_slice"))
        .where(...)
        .fetch();
```
Which results in `select entity_slice.entityName from (select * from entity_table for system_time as of transaction 12345) entity_slice` query and fails since column name should be `entity_slice.entity_name` instead of `entity_slice.entityName`.
This happens since meta models are generated from entity definitions using JPAAnnotationProcessor since I don't have DB during build time to generate model from (as documentation suggests for native sql compatible meta models) and in general would prefer to stick to JPA. I wasn't able to find any viable solution on how to embed native sql snippet into JPAQuery either.

Any suggestions on how one could facilitate system-versioned tables using querydsl?

**Comment by jwgmeligmeyling** [comment]
You should look into `QueryFlag` and `JoinFlag` 😄

**Comment by gnom7** [comment]
@jwgmeligmeyling thanks for your answer :+1: 

Are you proposing what is described [here](http://www.querydsl.com/static/querydsl/4.1.1/reference/html_single/#d0e1264)?
The only mention of `QueryFlag` from reference doc I could find.

Am I right that this requires introduction of completely new query type?
Which will require:
- to use different query types across application (existing ones and customized)
- leave user without out-of-the-box support of factories etc.
- require to implement 6 unrelated to problem methods from abstract class (`clone`, `createSerializer`, `select` x2, `iterate`, `fetchResults`)

Also I'm not sure about proper `Position`, i.e. in case I use `AFTER_SELECT` to create custom `FROM` would it cover all possible cases? I'm not sure.
Also chapter 2.3 is about `SQL` and thus `JpaQuery` doesn't support `addFlag` method and thus sql-invalid (jpa-valid) column `entity_slice.entityName` in the created query persists.

I'm quite new to querydsl, could you please expand on your question in case I miss something?

**Comment by jwgmeligmeyling** [comment]
No you won't need a new query type. The `BEFORE_FILTERS` are rendered just after the FROM / JOIN clauses.

See the following example:

```java
SQLQuery<?> query = select(survey.id).from(survey);
query.addFlag(QueryFlag.Position.BEFORE_FILTERS, Expressions.stringTemplate(" for system_time as of transaction {0} ", 12345l));
String s = query.toString();
```

Results in:

```
select SURVEY.ID
from SURVEY SURVEY for system_time as of transaction ?
```

**Comment by gnom7** [comment]
Appreciate your time @jwgmeligmeyling , thank you, going to close this
